### PR TITLE
prov/verbs: Prevent CQ and SQ overrun

### PR DESCRIPTION
**Pull request by shefty**

### CODE REVIEW

File: prov/verbs/src/verbs_ep.c [hunk 1]
```diff
@@ -36,6 +36,57 @@
...
+
+	ctx->ep = ep;
+	ctx->user_ctx = (void *) (uintptr_t) wr->wr_id;
+	wr->wr_id = (uintptr_t) ctx;
+
+	ret = ibv_post_send(ep->ibv_qp, wr, &bad_wr);
+	if (ret) {
+		cq->util_cq.cq_fastlock_acquire(&cq->util_cq.cq_lock);
+		cq->credits++;
+		goto free_cq_space;
```
Comment (swelch) [inline prov/verbs/src/verbs_ep.c]: Looks like a missing close parentheses got dropped.

File: prov/verbs/src/verbs_cq.c [hunk 2]
```diff
@@ -663,14 +639,18 @@ int fi_ibv_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
 		goto err6;
 	}
 
-	slist_init(&cq->wcq);
+	ret = ofi_bufpool_create(&cq->ctx_pool, sizeof(struct fi_context), 0,
+				 size, fi_ibv_gl_data.def_tx_size, 0);
+	if (ret)
```
Comment (swelch) [inline prov/verbs/src/verbs_cq.c]: Looks like size and alignment parameters are reversed.

File: prov/verbs/src/verbs_ep.c (hunk 1 above)
Comment (shefty) [reply prov/verbs/src/verbs_ep.c]: How did this compile cleanly?  Odd...

**Comment by shefty** [comment]
@swelch - Can I ask you to check xrc with these changes? These are now at least passing my internal tests.

**Comment by swelch** [comment]
> @swelch - Can I ask you to check xrc with these changes? These are now at least passing my internal tests.

I will take a look, thanks.

**Comment by jswaro** [comment]
@shefty You can add the cray-jenkins-build label and it should run our tests on every new commit.

**Comment by shefty** [comment]
@jswaro - thanks

**Comment by shefty** [comment]
Did adding the label verify that xrc still works, or should I wait on other testing for that?

**Comment by jswaro** [comment]
The last job ran 2 hours ago, and was successful.

**Comment by swelch** [comment]
@shefty - I've been doing a review as time permits and also had fetched the PR and have been testing. I've seen no issues in my testing (or what I've reviewed so far), but was going to run some very large scale tests in the next couple of hours and try to complete the review.

**Pull request merged**